File: src/main/java/studio/dreamys/prometheus/essential/util/OS.java
package studio.dreamys.prometheus.essential.util;

import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;

import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Locale;

public enum OS {
    Windows(Paths.get(".")),
    MacOS(PathUtil.fromHomeFolder("Library", "Application Support")),
    Linux(PathUtil.fromEnvOr("XDG_DATA_HOME", PathUtil.fromHomeFolder(".local", "share")))
    ;

    OS(@NotNull Path configFolder) {
        this.configFolder = configFolder;
    }
    private final @NotNull Path configFolder;
    private static final @NotNull OS current;
    static {
        String osName = System.getProperty("os.name").toLowerCase(Locale.ROOT);
        if (osName.contains("win")) {
            current = Windows;
        } else if (osName.contains("darwin") || osName.contains("mac")) {
            current = MacOS;
        } else {
            current = Linux;
        }
    }

    @Contract(value = "-> new", pure = true)
    public static @NotNull Path getConfigFolder() {
        return PathUtil.fromEnvOr("PROMETHEUS_FOLDER", current.configFolder.resolve("prometheus"));
    }

    @Contract(pure = true)
    public static boolean isOnWindows() {
        return current == Windows;
    }

    @Contract(pure = true)
    public static @NotNull String getName() {
        return current.toString();
    }
}

File: src/main/java/studio/dreamys/prometheus/essential/util/OS.kt
package studio.dreamys.prometheus.essential.util

import java.nio.file.Path
import kotlin.io.path.Path
import studio.dreamys.prometheus.essential.ext.*

enum class OS(val configFolder: Path) {
    Windows(Path(".")),
    // ~/.local/share is $XDG_DATA_HOME
    Linux(if (System.getenv("XDG_DATA_HOME").isNullOrBlank()) {
            Path(System.getenv("XDG_DATA_HOME")) 
        } else {
            Path(System.getProperty("user.home"), ".local", "share")
        }
    ),
    MacOS(Path(System.getProperty("user.home"), "Library", "Application Support"));
    companion object {
        val current: OS = run {
            val name = System.getProperty("os.name").lowercase()
            return@run if ("win" in name) {
                Windows
            } else if ("mac" in name || "darwin" in name) {
                MacOS
            } else {
                Linux
            }
        }
        fun isOnWindows(): Boolean {
            return current == Windows
        }
    }
}